mujoco_py/mjpid.pyx

```diff
@@ -80,21 +81,114 @@ cdef mjtNum c_pid_bias(const mjModel* m, const mjData* d, int id) with gil:
     d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_LAST_ERROR] = error
     d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_DERIVATIVE_ERROR_LAST] = derivative_error
     d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_INTEGRAL_ERROR] = integral_error
+
+    if effort_limit_low != 0.0 or effort_limit_high != 0.0:
+        f = fmax(effort_limit_low, fmin(effort_limit_high, f))
+    return f
+
+""" 
+    Inverse Dynamics (ID) Controller
     
+    qacc:               Joint acceleration.
+    qfrc_applied:       Torques applied directly to the joints.
+    xfrc_applied:       Cartesian forces applied directly to bodies.
+    qfrc_actuator:      Torques applied directly to the actuators.
+    Jx'*xfrc_applied:   Joint torque resulting from cartesian forces (xfrc_applied).
+    
+    qfrc_inverse gives the joint torques necessary to achieve a desired joint acceleration (qacc) given 
+    internal and external forces and torques. ID control solves the following torque balance by calling
+    mjinverse(model, data):
+    
+       qfrc_inverse = qfrc_applied + Jx'*xfrc_applied + qfrc_actuator
+    The error in desired joint acceleration is wrapped using a PD controller.
+    To provide a smooth reference signal for the ID controller, an Exponential Moving Average (EMA) is
+    used on the reference control signal (ctrl_ema). 
+"""
+cdef enum USER_DEFINED_INV_DYN_ACTUATOR_PARAMS:
+    IDX_INV_DYN_PROPORTIONAL_GAIN = 0,  # Kp, proportional gain
+    IDX_INV_DYN_DERIVATIVE_GAIN = 1,  #Kd, derivative gain
+    IDX_INV_DYN_EMA_SMOOTH = 2,  #Exponential moving average smoothing factor
+
+cdef enum USER_DEFINED_INV_DYN_CONTROLLER_DATA:
+    IDX_INV_DYN_CTRL_REF = 0  # stored value of EMA-smoothed control
+
+cdef mjtNum c_inv_dyn_bias(const mjModel*m, const mjData*d, int id):
+    cdef int NGAIN = int(const.NGAIN)
+    ctrl_ema = d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_INV_DYN_CTRL_REF]
+    ema_smooth = m.actuator_gainprm[id * NGAIN + IDX_INV_DYN_EMA_SMOOTH]
+    cdef double effort_limit_low = m.actuator_forcerange[id * 2]
+    cdef double effort_limit_high = m.actuator_forcerange[id * 2 + 1]
+
+    # Apply an Exponential Moving Average (EMA) to desired control
+    ctrl_ema = (ema_smooth * ctrl_ema) + (1 - ema_smooth) * d.ctrl[id]
+
+    qpos_des = ctrl_ema
+    qvel_des = 0
+
+    qpos_error = qpos_des - d.qpos[id]
+    qvel_error = qvel_des - d.qvel[id]
+
+    # PD gains for desired acceleration
+    kp = m.actuator_gainprm[id * NGAIN + IDX_INV_DYN_PROPORTIONAL_GAIN]
+    kd = m.actuator_gainprm[id * NGAIN + IDX_INV_DYN_DERIVATIVE_GAIN]
+
+    # Set desired acceleration of all DoFs (model.nv) to zero except the target actuator [id]
+    qacc_des = np.zeros(m.nv)
+    qacc_des[id] = kp * qpos_error + kd * qvel_error
+
+    # Set the target forward dyanmics
+    for i in range(m.nv):
+        d.qacc[i] = qacc_des[i]
```

Comment:
this seems strange, as each joint (specified by id) will be called on this function, so it will use mj_inverse for each joint independently, I guess mj_inverse is an global operation, it is in-efficient perhaps also introduce non-deterministic behavior  ; also this controller has two input and single output, is it stable? we shall add more tests.

Reply:
Ideally we do just as you say and only call mj_inverse once. However, this would require us to know how many actuators are using InvDyn and call mj_inverse on the last joint. This might not be robust in a mixture of PID and InvDyn actuators. 
With regards to the two inputs, are you talking about the PD terms on acceleration?

Reply:
yes, input is both pos error and velocity (0-v)

Reply:
how many inv_dyn joints do you plan to have?? it might worth making one joint special (e.g. added annotation into actuator_user), which responsible for doing this logic once for all joints once each step.